Unit tests of NailgunTasks flake in CI under the v2 pytest rule

Failures look like so:
```
tests/python/pants_test/backend/jvm/tasks:jar_task stdout:
============================= test session starts ==============================
platform linux -- Python 3.6.7, pytest-3.6.4, py-1.8.0, pluggy-0.7.1
rootdir: /home/travis/build/pantsbuild/pants/.pants.d/process-executionx72EnL, inifile:
plugins: cov-2.4.0, timeout-1.2.1
collected 13 items

pants_test/backend/jvm/tasks/test_jar_task.py ........F..F.              [100%]

=================================== FAILURES ===================================
_________________________ JarTaskTest.test_update_jars _________________________

self = NailgunClient(host='127.0.0.1', port=43797, workdir='/tmp/tmpual1wsud_BUILD_ROOT')

    def try_connect(self):
      """Creates a socket, connects it to the nailgun and returns the connected socket.
    
        :returns: a connected `socket.socket`.
        :raises: `NailgunClient.NailgunConnectionError` on failure to connect.
        """
      sock = RecvBufferedSocket(socket.socket(socket.AF_INET, socket.SOCK_STREAM))
      try:
>       sock.connect(self._address)
E       ConnectionRefusedError: [Errno 111] Connection refused

pants/java/nailgun_client.py:297: ConnectionRefusedError

During handling of the above exception, another exception occurred:

this = <pants.java.nailgun_executor.NailgunExecutor._runner.<locals>.Runner object at 0x7bf457a622e8>
stdout = <_pytest.capture.EncodedFile object at 0x7f27af379a20>
stderr = <_pytest.capture.EncodedFile object at 0x7f27af379a20>, stdin = None
cwd = None

    def run(this, stdout=None, stderr=None, stdin=None, cwd=None):
      nailgun = None
      try:
>       nailgun = self._get_nailgun_client(jvm_options, classpath, stdout, stderr, stdin)

pants/java/nailgun_executor.py:146: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

self = <pants.java.nailgun_executor.NailgunExecutor object at 0x7bf457a57da0>
jvm_options = ['-Xmx256m']
classpath = ['/tmp/tmpual1wsud_BUILD_ROOT/.pants.d/bootstrap_jvm_tools/7c0a62077e8f/ivy/jars/com.martiansoftware/nailgun-server/ja...tmpual1wsud_BUILD_ROOT/.pants.d/bootstrap_jvm_tools/7c0a62077e8f/ivy/jars/org.pantsbuild/args4j/jars/args4j-0.0.9.jar']
stdout = <_pytest.capture.EncodedFile object at 0x7f27af379a20>
stderr = <_pytest.capture.EncodedFile object at 0x7f27af379a20>, stdin = None

    def _get_nailgun_client(self, jvm_options, classpath, stdout, stderr, stdin):
      """This (somewhat unfortunately) is the main entrypoint to this class via the Runner. It handles
           creation of the running nailgun server as well as creation of the client."""
      classpath = self._nailgun_classpath + classpath
      new_fingerprint = self._fingerprint(jvm_options, classpath, self._distribution.version)
    
      with self._NAILGUN_SPAWN_LOCK:
        running, updated = self._check_nailgun_state(new_fingerprint)
    
        if running and updated:
          logger.debug('Found running nailgun server that needs updating, killing {server}'
                       .format(server=self._identity))
          self.terminate()
    
        if (not running) or (running and updated):
>         return self._spawn_nailgun_server(new_fingerprint, jvm_options, classpath, stdout, stderr, stdin)

pants/java/nailgun_executor.py:184: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

self = <pants.java.nailgun_executor.NailgunExecutor object at 0x7bf457a57da0>
fingerprint = '0103cbb4f4d354f0ff2c056aeb216910b18e12fd'
jvm_options = ['-Xmx256m', '-Dpants.buildroot=/home/travis/build/pantsbuild/pants', '-Dpants.nailgun.owner=/tmp/tmpual1wsud_BUILD_ROOT/.pants.d/ng/test_TestJarTask_test_scope', '-Dpants.nailgun.fingerprint=0103cbb4f4d354f0ff2c056aeb216910b18e12fd']
classpath = ['/tmp/tmpual1wsud_BUILD_ROOT/.pants.d/bootstrap_jvm_tools/7c0a62077e8f/ivy/jars/com.martiansoftware/nailgun-server/ja...tmpual1wsud_BUILD_ROOT/.pants.d/bootstrap_jvm_tools/7c0a62077e8f/ivy/jars/org.pantsbuild/args4j/jars/args4j-0.0.9.jar']
stdout = <_pytest.capture.EncodedFile object at 0x7f27af379a20>
stderr = <_pytest.capture.EncodedFile object at 0x7f27af379a20>, stdin = None

    def _spawn_nailgun_server(self, fingerprint, jvm_options, classpath, stdout, stderr, stdin):
      """Synchronously spawn a new nailgun server."""
      # Truncate the nailguns stdout & stderr.
      safe_file_dump(self._ng_stdout, b'', mode='wb')
      safe_file_dump(self._ng_stderr, b'', mode='wb')
    
      jvm_options = jvm_options + [self._PANTS_NG_BUILDROOT_ARG,
                                   self._create_owner_arg(self._workdir),
                                   self._create_fingerprint_arg(fingerprint)]
    
      post_fork_child_opts = dict(fingerprint=fingerprint,
                                  jvm_options=jvm_options,
                                  classpath=classpath,
                                  stdout=stdout,
                                  stderr=stderr)
    
      logger.debug('Spawning nailgun server {i} with fingerprint={f}, jvm_options={j}, classpath={cp}'
                   .format(i=self._identity, f=fingerprint, j=jvm_options, cp=classpath))
    
      self.daemon_spawn(post_fork_child_opts=post_fork_child_opts)
    
      # Wait for and write the port information in the parent so we can bail on exception/timeout.
      self.await_pid(self._startup_timeout)
      self.write_socket(self._await_socket(self._connect_timeout))
    
      logger.debug('Spawned nailgun server {i} with fingerprint={f}, pid={pid} port={port}'
                   .format(i=self._identity, f=fingerprint, pid=self.pid, port=self.socket))
    
      client = self._create_ngclient(self.socket, stdout, stderr, stdin)
>     self.ensure_connectable(client)

pants/java/nailgun_executor.py:273: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

self = <pants.java.nailgun_executor.NailgunExecutor object at 0x7bf457a57da0>
nailgun = NailgunClient(host='127.0.0.1', port=43797, workdir='/tmp/tmpual1wsud_BUILD_ROOT')

    def ensure_connectable(self, nailgun):
      """Ensures that a nailgun client is connectable or raises NailgunError."""
      attempt_count = 1
      while 1:
        try:
>         with closing(nailgun.try_connect()) as sock:

pants/java/nailgun_executor.py:233: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

self = NailgunClient(host='127.0.0.1', port=43797, workdir='/tmp/tmpual1wsud_BUILD_ROOT')

    def try_connect(self):
      """Creates a socket, connects it to the nailgun and returns the connected socket.
    
        :returns: a connected `socket.socket`.
        :raises: `NailgunClient.NailgunConnectionError` on failure to connect.
        """
      sock = RecvBufferedSocket(socket.socket(socket.AF_INET, socket.SOCK_STREAM))
      try:
        sock.connect(self._address)
      except (socket.error, socket.gaierror) as e:
        logger.debug('Encountered socket exception {!r} when attempting connect to nailgun'.format(e))
        sock.close()
        raise self.NailgunConnectionError(
          address=self._address_string,
          pid=self._maybe_last_pid(),
          pgrp=self._maybe_last_pgrp(),
>         wrapped_exc=e,
        )
E       pants.java.nailgun_client.NailgunClient.NailgunConnectionError: ("Problem connecting to nailgun server (address: 127.0.0.1:43797, remote_pid=<remote PID chunk not yet received!>, remote_pgrp=<remote PGRP chunk not yet received!>): ConnectionRefusedError(111, 'Connection refused')", ConnectionRefusedError(111, 'Connection refused'))

pants/java/nailgun_client.py:305: NailgunConnectionError

During handling of the above exception, another exception occurred:

self = <abc.test_TestJarTask_test_scope object at 0x7bf457ac61d0>
classpath = ['/tmp/tmpual1wsud_BUILD_ROOT/.pants.d/bootstrap_jvm_tools/7c0a62077e8f/ivy/jars/org.pantsbuild/jar-tool/jars/jar-tool...tmpual1wsud_BUILD_ROOT/.pants.d/bootstrap_jvm_tools/7c0a62077e8f/ivy/jars/org.pantsbuild/args4j/jars/args4j-0.0.9.jar']
main = 'org.pantsbuild.tools.jar.Main', jvm_options = ['-Xmx256m']
args = ['-files=/tmp/tmpapcvgqvw/tmpacza3pbe=a/b', '-update=true', '-compress=true', '-default_action=SKIP', '-skip=^META-INF...^META-INF/[^/]+\\.DSA$,^META-INF/[^/]+\\.RSA$,^META-INF/INDEX.LIST$', '-policies=^META-INF/services/=CONCAT_TEXT', ...]
workunit_name = 'jar-tool'
workunit_labels = ['TOOL', 'JVM', 'NAILGUN', 'SUPPRESS_LABEL']
workunit_log_config = None, dist = None

    def runjava(self, classpath, main, jvm_options=None, args=None, workunit_name=None,
                workunit_labels=None, workunit_log_config=None, dist=None):
      """Runs the java main using the given classpath and args.
    
        If --execution-strategy=subprocess is specified then the java main is run in a freshly spawned
        subprocess, otherwise a persistent nailgun server dedicated to this Task subclass is used to
        speed up amortized run times.
    
        :API: public
        """
      executor = self.create_java_executor(dist=dist)
    
      # Creating synthetic jar to work around system arg length limit is not necessary
      # when `NailgunExecutor` is used because args are passed through socket, therefore turning off
      # creating synthetic jar if nailgun is used.
      create_synthetic_jar = self.execution_strategy != self.NAILGUN
      try:
        return util.execute_java(classpath=classpath,
                                 main=main,
                                 jvm_options=jvm_options,
                                 args=args,
                                 executor=executor,
                                 workunit_factory=self.context.new_workunit,
                                 workunit_name=workunit_name,
                                 workunit_labels=workunit_labels,
                                 workunit_log_config=workunit_log_config,
                                 create_synthetic_jar=create_synthetic_jar,
>                                synthetic_jar_dir=self._executor_workdir)

pants/backend/jvm/tasks/nailgun_task.py:121: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

classpath = ['/tmp/tmpual1wsud_BUILD_ROOT/.pants.d/bootstrap_jvm_tools/7c0a62077e8f/ivy/jars/org.pantsbuild/jar-tool/jars/jar-tool...tmpual1wsud_BUILD_ROOT/.pants.d/bootstrap_jvm_tools/7c0a62077e8f/ivy/jars/org.pantsbuild/args4j/jars/args4j-0.0.9.jar']
main = 'org.pantsbuild.tools.jar.Main', jvm_options = ['-Xmx256m']
args = ['-files=/tmp/tmpapcvgqvw/tmpacza3pbe=a/b', '-update=true', '-compress=true', '-default_action=SKIP', '-skip=^META-INF...^META-INF/[^/]+\\.DSA$,^META-INF/[^/]+\\.RSA$,^META-INF/INDEX.LIST$', '-policies=^META-INF/services/=CONCAT_TEXT', ...]
executor = <pants.java.nailgun_executor.NailgunExecutor object at 0x7bf457a57da0>
workunit_factory = <bound method TestContext.new_workunit of <pants_test.base.context_utils.TestContext object at 0x7bf457abafd0>>
workunit_name = 'jar-tool'
workunit_labels = ['TOOL', 'JVM', 'NAILGUN', 'SUPPRESS_LABEL'], cwd = None
workunit_log_config = None, distribution = None, create_synthetic_jar = False
synthetic_jar_dir = '/tmp/tmpual1wsud_BUILD_ROOT/.pants.d/ng/test_TestJarTask_test_scope'
stdin = None

    def execute_java(classpath, main, jvm_options=None, args=None, executor=None,
                     workunit_factory=None, workunit_name=None, workunit_labels=None,
                     cwd=None, workunit_log_config=None, distribution=None,
                     create_synthetic_jar=True, synthetic_jar_dir=None, stdin=None):
      """Executes the java program defined by the classpath and main.
    
      If `workunit_factory` is supplied, does so in the context of a workunit.
    
      :param list classpath: the classpath for the java program
      :param string main: the fully qualified class name of the java program's entry point
      :param list jvm_options: an optional sequence of options for the underlying jvm
      :param list args: an optional sequence of args to pass to the java program
      :param executor: an optional java executor to use to launch the program; defaults to a subprocess
        spawn of the default java distribution
      :param workunit_factory: an optional callable that can produce a workunit context
      :param string workunit_name: an optional name for the work unit; defaults to the main
      :param list workunit_labels: an optional sequence of labels for the work unit
      :param string cwd: optionally set the working directory
      :param WorkUnit.LogConfig workunit_log_config: an optional tuple of options affecting reporting
      :param bool create_synthetic_jar: whether to create a synthentic jar that includes the original
        classpath in its manifest.
      :param string synthetic_jar_dir: an optional directory to store the synthetic jar, if `None`
        a temporary directory will be provided and cleaned up upon process exit.
      :param file stdin: The stdin handle to use: by default None, meaning that stdin will
        not be propagated into the process.
    
      Returns the exit code of the java program.
      Raises `pants.java.Executor.Error` if there was a problem launching java itself.
      """
    
      runner = _get_runner(classpath, main, jvm_options, args, executor, cwd, distribution,
                           create_synthetic_jar, synthetic_jar_dir)
      workunit_name = workunit_name or main
    
      return execute_runner(runner,
                            workunit_factory=workunit_factory,
                            workunit_name=workunit_name,
                            workunit_labels=workunit_labels,
                            workunit_log_config=workunit_log_config,
>                           stdin=stdin)

pants/java/util.py:80: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

runner = <pants.java.nailgun_executor.NailgunExecutor._runner.<locals>.Runner object at 0x7bf457a622e8>
workunit_factory = <bound method TestContext.new_workunit of <pants_test.base.context_utils.TestContext object at 0x7bf457abafd0>>
workunit_name = 'jar-tool'
workunit_labels = ['TOOL', 'NAILGUN', 'TOOL', 'JVM', 'NAILGUN', 'SUPPRESS_LABEL']
workunit_log_config = None, stdin = None

    def execute_runner(runner, workunit_factory=None, workunit_name=None, workunit_labels=None,
                       workunit_log_config=None, stdin=None):
      """Executes the given java runner.
    
      If `workunit_factory` is supplied, does so in the context of a workunit.
    
      :param runner: the java runner to run
      :param workunit_factory: an optional callable that can produce a workunit context
      :param string workunit_name: an optional name for the work unit; defaults to the main
      :param list workunit_labels: an optional sequence of labels for the work unit
      :param WorkUnit.LogConfig workunit_log_config: an optional tuple of task options affecting reporting
      :param file stdin: The stdin handle to use: by default None, meaning that stdin will
        not be propagated into the process.
    
      Returns the exit code of the java runner.
      Raises `pants.java.Executor.Error` if there was a problem launching java itself.
      """
      if not isinstance(runner, Executor.Runner):
        raise ValueError('The runner argument must be a java Executor.Runner instance, '
                         'given {} of type {}'.format(runner, type(runner)))
    
      if workunit_factory is None:
        return runner.run(stdin=stdin)
      else:
        workunit_labels = [
            WorkUnitLabel.TOOL,
            WorkUnitLabel.NAILGUN if isinstance(runner.executor, NailgunExecutor) else WorkUnitLabel.JVM
        ] + (workunit_labels or [])
    
        with workunit_factory(name=workunit_name, labels=workunit_labels,
                              cmd=runner.cmd, log_config=workunit_log_config) as workunit:
          ret = runner.run(stdout=workunit.output('stdout'),
                           stderr=workunit.output('stderr'),
>                          stdin=stdin)

pants/java/util.py:156: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

this = <pants.java.nailgun_executor.NailgunExecutor._runner.<locals>.Runner object at 0x7bf457a622e8>
stdout = <_pytest.capture.EncodedFile object at 0x7f27af379a20>
stderr = <_pytest.capture.EncodedFile object at 0x7f27af379a20>, stdin = None
cwd = None

    def run(this, stdout=None, stderr=None, stdin=None, cwd=None):
      nailgun = None
      try:
        nailgun = self._get_nailgun_client(jvm_options, classpath, stdout, stderr, stdin)
        logger.debug('Executing via {ng_desc}: {cmd}'.format(ng_desc=nailgun, cmd=this.cmd))
        return nailgun.execute(main, cwd, *args)
      except (NailgunClient.NailgunError, self.InitialNailgunConnectTimedOut) as e:
        self.terminate()
        raise self.Error('Problem launching via {ng_desc} command {main} {args}: {msg}'
                         .format(ng_desc=nailgun or '<no nailgun connection>',
                                 main=main,
                                 args=' '.join(args),
>                                msg=e))
E       pants.java.executor.Executor.Error: Problem launching via <no nailgun connection> command org.pantsbuild.tools.jar.Main -files=/tmp/tmpapcvgqvw/tmpacza3pbe=a/b -update=true -compress=true -default_action=SKIP -skip=^META-INF/[^/]+\.SF$,^META-INF/[^/]+\.DSA$,^META-INF/[^/]+\.RSA$,^META-INF/INDEX.LIST$ -policies=^META-INF/services/=CONCAT_TEXT /tmp/tmp9ddkgd84/tmp7lui29_b.jar: ("Problem connecting to nailgun server (address: 127.0.0.1:43797, remote_pid=<remote PID chunk not yet received!>, remote_pgrp=<remote PGRP chunk not yet received!>): ConnectionRefusedError(111, 'Connection refused')", ConnectionRefusedError(111, 'Connection refused'))

pants/java/nailgun_executor.py:155: Error

During handling of the above exception, another exception occurred:

self = <pants_test.backend.jvm.tasks.test_jar_task.JarTaskTest testMethod=test_update_jars>

    def test_update_jars(self):
      with self.jarfile() as main_jar:
        with self.jarfile() as included_jar:
          with self.jar_task.open_jar(main_jar) as jar:
>           jar.writestr('a/b', b'c')

pants_test/backend/jvm/tasks/test_jar_task.py:203: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
/opt/pyenv/versions/3.6.7/lib/python3.6/contextlib.py:88: in __exit__
    next(self.gen)
pants/backend/jvm/tasks/jar_task.py:360: in open_jar
    if JarTool.global_instance().run(context=self.context, runjava=self.runjava, args=args):
pants/backend/jvm/subsystems/jar_tool.py:33: in run
    WorkUnitLabel.NAILGUN, WorkUnitLabel.SUPPRESS_LABEL])
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

self = <abc.test_TestJarTask_test_scope object at 0x7bf457ac61d0>
classpath = ['/tmp/tmpual1wsud_BUILD_ROOT/.pants.d/bootstrap_jvm_tools/7c0a62077e8f/ivy/jars/org.pantsbuild/jar-tool/jars/jar-tool...tmpual1wsud_BUILD_ROOT/.pants.d/bootstrap_jvm_tools/7c0a62077e8f/ivy/jars/org.pantsbuild/args4j/jars/args4j-0.0.9.jar']
main = 'org.pantsbuild.tools.jar.Main', jvm_options = ['-Xmx256m']
args = ['-files=/tmp/tmpapcvgqvw/tmpacza3pbe=a/b', '-update=true', '-compress=true', '-default_action=SKIP', '-skip=^META-INF...^META-INF/[^/]+\\.DSA$,^META-INF/[^/]+\\.RSA$,^META-INF/INDEX.LIST$', '-policies=^META-INF/services/=CONCAT_TEXT', ...]
workunit_name = 'jar-tool'
workunit_labels = ['TOOL', 'JVM', 'NAILGUN', 'SUPPRESS_LABEL']
workunit_log_config = None, dist = None

    def runjava(self, classpath, main, jvm_options=None, args=None, workunit_name=None,
                workunit_labels=None, workunit_log_config=None, dist=None):
      """Runs the java main using the given classpath and args.
    
        If --execution-strategy=subprocess is specified then the java main is run in a freshly spawned
        subprocess, otherwise a persistent nailgun server dedicated to this Task subclass is used to
        speed up amortized run times.
    
        :API: public
        """
      executor = self.create_java_executor(dist=dist)
    
      # Creating synthetic jar to work around system arg length limit is not necessary
      # when `NailgunExecutor` is used because args are passed through socket, therefore turning off
      # creating synthetic jar if nailgun is used.
      create_synthetic_jar = self.execution_strategy != self.NAILGUN
      try:
        return util.execute_java(classpath=classpath,
                                 main=main,
                                 jvm_options=jvm_options,
                                 args=args,
                                 executor=executor,
                                 workunit_factory=self.context.new_workunit,
                                 workunit_name=workunit_name,
                                 workunit_labels=workunit_labels,
                                 workunit_log_config=workunit_log_config,
                                 create_synthetic_jar=create_synthetic_jar,
                                 synthetic_jar_dir=self._executor_workdir)
      except executor.Error as e:
>       raise TaskError(e)
E       pants.base.exceptions.TaskError: Problem launching via <no nailgun connection> command org.pantsbuild.tools.jar.Main -files=/tmp/tmpapcvgqvw/tmpacza3pbe=a/b -update=true -compress=true -default_action=SKIP -skip=^META-INF/[^/]+\.SF$,^META-INF/[^/]+\.DSA$,^META-INF/[^/]+\.RSA$,^META-INF/INDEX.LIST$ -policies=^META-INF/services/=CONCAT_TEXT /tmp/tmp9ddkgd84/tmp7lui29_b.jar: ("Problem connecting to nailgun server (address: 127.0.0.1:43797, remote_pid=<remote PID chunk not yet received!>, remote_pgrp=<remote PGRP chunk not yet received!>): ConnectionRefusedError(111, 'Connection refused')", ConnectionRefusedError(111, 'Connection refused'))

pants/backend/jvm/tasks/nailgun_task.py:123: TaskError
----------------------------- Captured stderr call -----------------------------

Starting workunit cache
.
Starting workunit bootstrap-jar-tool
:: loading settings :: url = jar:file:/home/travis/.ivy2/pants/org.apache.ivy/ivy/jars/ivy-2.4.0.jar!/org/apache/ivy/core/settings/ivysettings.xml
:: resolving dependencies :: internal#42f1fedb02831e5a4cb63034fb17bbb5d4a52248-IvyResolveFingerprintStrategy_76d7a3f2dc89-fetch;working@travis-job-2321f98e-28d0-40b4-844c-429f90eba81f
	confs: [default]
	found org.pantsbuild#jar-tool;0.0.10 in default
	found args4j#args4j;2.32 in default
	found com.google.code.findbugs#jsr305;3.0.0 in default
	found com.google.guava#guava;18.0 in default
	found org.pantsbuild#args4j;0.0.9 in default
:: resolution report :: resolve 370ms :: artifacts dl 24ms
	---------------------------------------------------------------------
	|                  |            modules            ||   artifacts   |
	|       conf       | number| search|dwnlded|evicted|| number|dwnlded|
	---------------------------------------------------------------------
	|      default     |   5   |   0   |   0   |   0   ||   5   |   0   |
	---------------------------------------------------------------------
cachepath output to /tmp/tmpual1wsud_BUILD_ROOT/.pants.d/bootstrap_jvm_tools/7c0a62077e8f/ivy/42f1fedb02831e5a4cb63034fb17bbb5d4a52248-IvyResolveFingerprintStrategy_76d7a3f2dc89/classpath.raw.tmp.bf687fd95d544e52958642ff85c42aa0

WorkUnit outcome: SUCCESS

Starting workunit cache
.
Starting workunit bootstrap-nailgun-server
:: loading settings :: url = jar:file:/home/travis/.ivy2/pants/org.apache.ivy/ivy/jars/ivy-2.4.0.jar!/org/apache/ivy/core/settings/ivysettings.xml
:: resolving dependencies :: internal#c109bfa74ccd26cbc0a32b4145338ab536124702-IvyResolveFingerprintStrategy_76d7a3f2dc89-fetch;working@travis-job-2321f98e-28d0-40b4-844c-429f90eba81f
	confs: [default]
	found com.martiansoftware#nailgun-server;0.9.1 in default
:: resolution report :: resolve 135ms :: artifacts dl 3ms
	---------------------------------------------------------------------
	|                  |            modules            ||   artifacts   |
	|       conf       | number| search|dwnlded|evicted|| number|dwnlded|
	---------------------------------------------------------------------
	|      default     |   1   |   0   |   0   |   0   ||   1   |   0   |
	---------------------------------------------------------------------
cachepath output to /tmp/tmpual1wsud_BUILD_ROOT/.pants.d/bootstrap_jvm_tools/7c0a62077e8f/ivy/c109bfa74ccd26cbc0a32b4145338ab536124702-IvyResolveFingerprintStrategy_76d7a3f2dc89/classpath.raw.tmp.a50bc87d5fbd4e8c8422003fc3ee58a6

WorkUnit outcome: SUCCESS

Starting workunit jar-tool
______________________ JarBuilderTest.test_agent_manifest ______________________

cls = <class 'pants.java.nailgun_client.NailgunClientSession'>
sock = <pants.util.socket.RecvBufferedSocket object at 0x7bf4579d9da0>
return_bytes = True

    @classmethod
    def read_chunk(cls, sock, return_bytes=False):
      """Read a single chunk from the connected client.
    
         A "chunk" is a variable-length block of data beginning with a 5-byte chunk header and followed
         by an optional payload. The chunk header consists of:
    
              1) The length of the chunk's payload (not including the header) as a four-byte big-endian
                 unsigned long. The high-order byte is header[0] and the low-order byte is header[3].
    
              2) A single byte identifying the type of chunk.
        """
      try:
        # Read the chunk header from the socket.
>       header = cls._read_until(sock, cls.HEADER_BYTES)

pants/java/nailgun_protocol.py:199: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

cls = <class 'pants.java.nailgun_client.NailgunClientSession'>
sock = <pants.util.socket.RecvBufferedSocket object at 0x7bf4579d9da0>
desired_size = 5

    @classmethod
    def _read_until(cls, sock, desired_size):
      """Read a certain amount of content from a socket before returning."""
      buf = b''
      while len(buf) < desired_size:
        recv_bytes = sock.recv(desired_size - len(buf))
        if not recv_bytes:
          raise cls.TruncatedRead('Expected {} bytes before socket shutdown, instead received {}'
>                                 .format(desired_size, len(buf)))
E         pants.java.nailgun_protocol.NailgunProtocol.TruncatedRead: Expected 5 bytes before socket shutdown, instead received 0

pants/java/nailgun_protocol.py:181: TruncatedRead

During handling of the above exception, another exception occurred:

self = NailgunClient(host='127.0.0.1', port=40527, workdir='/tmp/tmpj6awopxx_BUILD_ROOT')
main_class = 'org.pantsbuild.tools.jar.Main'
cwd = '/tmp/tmpj6awopxx_BUILD_ROOT'
args = ('-manifest=/tmp/tmpvtfgz4gu/tmp50t1xzsa', '-files=/tmp/tmpj6awopxx_BUILD_ROOT/.pants.d/pants_test_backend_jvm_tasks_t...t_action=SKIP', '-skip=^META-INF/[^/]+\\.SF$,^META-INF/[^/]+\\.DSA$,^META-INF/[^/]+\\.RSA$,^META-INF/INDEX.LIST$', ...)
environment = {'NAILGUN_FILESEPARATOR': '/', 'NAILGUN_PATHSEPARATOR': ':'}
sock = <pants.util.socket.RecvBufferedSocket object at 0x7bf4579d9da0>

    def execute(self, main_class, cwd=None, *args, **environment):
      """Executes the given main_class with any supplied args in the given environment.
    
        :param string main_class: the fully qualified class name of the main entrypoint
        :param string cwd: Set the working directory for this command
        :param list args: any arguments to pass to the main entrypoint
        :param dict environment: an env mapping made available to native nails via the nail context
        :returns: the exit code of the main_class.
        :raises: :class:`NailgunClient.NailgunError` if there was an error during execution.
        """
      environment = dict(**environment)
      environment.update(self.ENV_DEFAULTS)
      cwd = cwd or self._workdir
    
      sock = self.try_connect()
    
      # TODO(#6579): NailgunClientSession currently requires callbacks because it can't depend on
      # having received these chunks, so we need to avoid clobbering these fields until we initialize
      # a new session.
      self._current_remote_pid = None
      self._current_remote_pgrp = None
      self._session = NailgunClientSession(
        sock=sock,
        in_file=self._stdin,
        out_file=self._stdout,
        err_file=self._stderr,
        exit_on_broken_pipe=self._exit_on_broken_pipe,
        remote_pid_callback=self._receive_remote_pid,
        remote_pgrp_callback=self._receive_remote_pgrp)
      try:
>       return self._session.execute(cwd, main_class, *args, **environment)

pants/java/nailgun_client.py:360: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

self = <pants.java.nailgun_client.NailgunClientSession object at 0x7bf4579d96d8>
working_dir = '/tmp/tmpj6awopxx_BUILD_ROOT'
main_class = 'org.pantsbuild.tools.jar.Main'
arguments = ('-manifest=/tmp/tmpvtfgz4gu/tmp50t1xzsa', '-files=/tmp/tmpj6awopxx_BUILD_ROOT/.pants.d/pants_test_backend_jvm_tasks_t...t_action=SKIP', '-skip=^META-INF/[^/]+\\.SF$,^META-INF/[^/]+\\.DSA$,^META-INF/[^/]+\\.RSA$,^META-INF/INDEX.LIST$', ...)
environment = {'NAILGUN_FILESEPARATOR': '/', 'NAILGUN_PATHSEPARATOR': ':'}

    def execute(self, working_dir, main_class, *arguments, **environment):
      # Send the nailgun request.
      self.send_request(self._sock, working_dir, main_class, *arguments, **environment)
    
      # Process the remainder of the nailgun session.
>     return self._process_session()

pants/java/nailgun_client.py:176: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

self = <pants.java.nailgun_client.NailgunClientSession object at 0x7bf4579d96d8>

    def _process_session(self):
      """Process the outputs of the nailgun session.
    
        :raises: :class:`NailgunProtocol.ProcessStreamTimeout` if a timeout set from a signal handler
                                                               with .set_exit_timeout() completes.
        :raises: :class:`Exception` if the session completes before the timeout, the `reason` argument
                                    to .set_exit_timeout() will be raised."""
      try:
        for chunk_type, payload in self.iter_chunks(
          MaybeShutdownSocket(self._sock),
          return_bytes=True,
>         timeout_object=self,
        ):

pants/java/nailgun_client.py:121: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

cls = <class 'pants.java.nailgun_client.NailgunClientSession'>
maybe_shutdown_socket = <pants.java.nailgun_protocol.MaybeShutdownSocket object at 0x7bf457c08d30>
return_bytes = True
timeout_object = <pants.java.nailgun_client.NailgunClientSession object at 0x7bf4579d96d8>

    @classmethod
    def iter_chunks(cls, maybe_shutdown_socket, return_bytes=False, timeout_object=None):
      """Generates chunks from a connected socket until an Exit chunk is sent or a timeout occurs.
    
        :param sock: the socket to read from.
        :param bool return_bytes: If False, decode the payload into a utf-8 string.
        :param cls.TimeoutProvider timeout_object: If provided, will be checked every iteration for a
                                                   possible timeout.
        :raises: :class:`cls.ProcessStreamTimeout`
        """
      assert(timeout_object is None or isinstance(timeout_object, cls.TimeoutProvider))
    
      if timeout_object is None:
        deadline = None
      else:
        options = timeout_object.maybe_timeout_options()
        if options is None:
          deadline = None
        else:
          deadline = options.start_time + options.interval
    
      while 1:
        if deadline is not None:
          overtime_seconds = deadline - time.time()
          if overtime_seconds > 0:
            original_timestamp = datetime.datetime.fromtimestamp(deadline).isoformat()
            raise cls.ProcessStreamTimeout(
              "iterating over bytes from nailgun timed out at {}, overtime seconds: {}"
              .format(original_timestamp, overtime_seconds))
    
        with maybe_shutdown_socket.lock:
          if maybe_shutdown_socket.is_shutdown:
            break
          # We poll with low timeouts because we poll under a lock. This allows the DaemonPantsRunner
          # to shut down the socket, and us to notice, pretty quickly.
          with cls._set_socket_timeout(maybe_shutdown_socket.socket, timeout=0.01):
            try:
>             chunk_type, payload = cls.read_chunk(maybe_shutdown_socket.socket, return_bytes)

pants/java/nailgun_protocol.py:291: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

cls = <class 'pants.java.nailgun_client.NailgunClientSession'>
sock = <pants.util.socket.RecvBufferedSocket object at 0x7bf4579d9da0>
return_bytes = True

    @classmethod
    def read_chunk(cls, sock, return_bytes=False):
      """Read a single chunk from the connected client.
    
         A "chunk" is a variable-length block of data beginning with a 5-byte chunk header and followed
         by an optional payload. The chunk header consists of:
    
              1) The length of the chunk's payload (not including the header) as a four-byte big-endian
                 unsigned long. The high-order byte is header[0] and the low-order byte is header[3].
    
              2) A single byte identifying the type of chunk.
        """
      try:
        # Read the chunk header from the socket.
        header = cls._read_until(sock, cls.HEADER_BYTES)
      except cls.TruncatedRead as e:
>       raise cls.TruncatedHeaderError('Failed to read nailgun chunk header ({!r}).'.format(e))
E       pants.java.nailgun_protocol.NailgunProtocol.TruncatedHeaderError: Failed to read nailgun chunk header (TruncatedRead('Expected 5 bytes before socket shutdown, instead received 0',)).

pants/java/nailgun_protocol.py:201: TruncatedHeaderError

During handling of the above exception, another exception occurred:

this = <pants.java.nailgun_executor.NailgunExecutor._runner.<locals>.Runner object at 0x7bf4579d97f0>
stdout = <_pytest.capture.EncodedFile object at 0x7f27af379a20>
stderr = <_pytest.capture.EncodedFile object at 0x7f27af379a20>, stdin = None
cwd = None

    def run(this, stdout=None, stderr=None, stdin=None, cwd=None):
      nailgun = None
      try:
        nailgun = self._get_nailgun_client(jvm_options, classpath, stdout, stderr, stdin)
        logger.debug('Executing via {ng_desc}: {cmd}'.format(ng_desc=nailgun, cmd=this.cmd))
>       return nailgun.execute(main, cwd, *args)

pants/java/nailgun_executor.py:148: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

self = NailgunClient(host='127.0.0.1', port=40527, workdir='/tmp/tmpj6awopxx_BUILD_ROOT')
main_class = 'org.pantsbuild.tools.jar.Main'
cwd = '/tmp/tmpj6awopxx_BUILD_ROOT'
args = ('-manifest=/tmp/tmpvtfgz4gu/tmp50t1xzsa', '-files=/tmp/tmpj6awopxx_BUILD_ROOT/.pants.d/pants_test_backend_jvm_tasks_t...t_action=SKIP', '-skip=^META-INF/[^/]+\\.SF$,^META-INF/[^/]+\\.DSA$,^META-INF/[^/]+\\.RSA$,^META-INF/INDEX.LIST$', ...)
environment = {'NAILGUN_FILESEPARATOR': '/', 'NAILGUN_PATHSEPARATOR': ':'}
sock = <pants.util.socket.RecvBufferedSocket object at 0x7bf4579d9da0>

    def execute(self, main_class, cwd=None, *args, **environment):
      """Executes the given main_class with any supplied args in the given environment.
    
        :param string main_class: the fully qualified class name of the main entrypoint
        :param string cwd: Set the working directory for this command
        :param list args: any arguments to pass to the main entrypoint
        :param dict environment: an env mapping made available to native nails via the nail context
        :returns: the exit code of the main_class.
        :raises: :class:`NailgunClient.NailgunError` if there was an error during execution.
        """
      environment = dict(**environment)
      environment.update(self.ENV_DEFAULTS)
      cwd = cwd or self._workdir
    
      sock = self.try_connect()
    
      # TODO(#6579): NailgunClientSession currently requires callbacks because it can't depend on
      # having received these chunks, so we need to avoid clobbering these fields until we initialize
      # a new session.
      self._current_remote_pid = None
      self._current_remote_pgrp = None
      self._session = NailgunClientSession(
        sock=sock,
        in_file=self._stdin,
        out_file=self._stdout,
        err_file=self._stderr,
        exit_on_broken_pipe=self._exit_on_broken_pipe,
        remote_pid_callback=self._receive_remote_pid,
        remote_pgrp_callback=self._receive_remote_pgrp)
      try:
        return self._session.execute(cwd, main_class, *args, **environment)
      except (socket.error, NailgunProtocol.ProtocolError) as e:
        raise self.NailgunError(
          address=self._address_string,
          pid=self._maybe_last_pid(),
          pgrp=self._maybe_last_pgrp(),
>         wrapped_exc=e,
        )
E       pants.java.nailgun_client.NailgunClient.NailgunError: ('Problem talking to nailgun server (address: 127.0.0.1:40527, remote_pid=<remote PID chunk not yet received!>, remote_pgrp=<remote PGRP chunk not yet received!>): TruncatedHeaderError("Failed to read nailgun chunk header (TruncatedRead(\'Expected 5 bytes before socket shutdown, instead received 0\',)).",)', TruncatedHeaderError("Failed to read nailgun chunk header (TruncatedRead('Expected 5 bytes before socket shutdown, instead received 0',)).",))

pants/java/nailgun_client.py:366: NailgunError

During handling of the above exception, another exception occurred:

self = <abc.test_TestJarBuilderTask_test_scope object at 0x7bf457bc9550>
classpath = ['/tmp/tmpj6awopxx_BUILD_ROOT/.pants.d/bootstrap_jvm_tools/7c0a62077e8f/ivy/jars/org.pantsbuild/jar-tool/jars/jar-tool...tmpj6awopxx_BUILD_ROOT/.pants.d/bootstrap_jvm_tools/7c0a62077e8f/ivy/jars/org.pantsbuild/args4j/jars/args4j-0.0.9.jar']
main = 'org.pantsbuild.tools.jar.Main', jvm_options = ['-Xmx256m']
args = ['-manifest=/tmp/tmpvtfgz4gu/tmp50t1xzsa', '-files=/tmp/tmpj6awopxx_BUILD_ROOT/.pants.d/pants_test_backend_jvm_tasks_t...t_action=SKIP', '-skip=^META-INF/[^/]+\\.SF$,^META-INF/[^/]+\\.DSA$,^META-INF/[^/]+\\.RSA$,^META-INF/INDEX.LIST$', ...]
workunit_name = 'jar-tool'
workunit_labels = ['TOOL', 'JVM', 'NAILGUN', 'SUPPRESS_LABEL']
workunit_log_config = None, dist = None

    def runjava(self, classpath, main, jvm_options=None, args=None, workunit_name=None,
                workunit_labels=None, workunit_log_config=None, dist=None):
      """Runs the java main using the given classpath and args.
    
        If --execution-strategy=subprocess is specified then the java main is run in a freshly spawned
        subprocess, otherwise a persistent nailgun server dedicated to this Task subclass is used to
        speed up amortized run times.
    
        :API: public
        """
      executor = self.create_java_executor(dist=dist)
    
      # Creating synthetic jar to work around system arg length limit is not necessary
      # when `NailgunExecutor` is used because args are passed through socket, therefore turning off
      # creating synthetic jar if nailgun is used.
      create_synthetic_jar = self.execution_strategy != self.NAILGUN
      try:
        return util.execute_java(classpath=classpath,
                                 main=main,
                                 jvm_options=jvm_options,
                                 args=args,
                                 executor=executor,
                                 workunit_factory=self.context.new_workunit,
                                 workunit_name=workunit_name,
                                 workunit_labels=workunit_labels,
                                 workunit_log_config=workunit_log_config,
                                 create_synthetic_jar=create_synthetic_jar,
>                                synthetic_jar_dir=self._executor_workdir)

pants/backend/jvm/tasks/nailgun_task.py:121: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

classpath = ['/tmp/tmpj6awopxx_BUILD_ROOT/.pants.d/bootstrap_jvm_tools/7c0a62077e8f/ivy/jars/org.pantsbuild/jar-tool/jars/jar-tool...tmpj6awopxx_BUILD_ROOT/.pants.d/bootstrap_jvm_tools/7c0a62077e8f/ivy/jars/org.pantsbuild/args4j/jars/args4j-0.0.9.jar']
main = 'org.pantsbuild.tools.jar.Main', jvm_options = ['-Xmx256m']
args = ['-manifest=/tmp/tmpvtfgz4gu/tmp50t1xzsa', '-files=/tmp/tmpj6awopxx_BUILD_ROOT/.pants.d/pants_test_backend_jvm_tasks_t...t_action=SKIP', '-skip=^META-INF/[^/]+\\.SF$,^META-INF/[^/]+\\.DSA$,^META-INF/[^/]+\\.RSA$,^META-INF/INDEX.LIST$', ...]
executor = <pants.java.nailgun_executor.NailgunExecutor object at 0x7bf457a315f8>
workunit_factory = <bound method TestContext.new_workunit of <pants_test.base.context_utils.TestContext object at 0x7bf457a26518>>
workunit_name = 'jar-tool'
workunit_labels = ['TOOL', 'JVM', 'NAILGUN', 'SUPPRESS_LABEL'], cwd = None
workunit_log_config = None, distribution = None, create_synthetic_jar = False
synthetic_jar_dir = '/tmp/tmpj6awopxx_BUILD_ROOT/.pants.d/ng/test_TestJarBuilderTask_test_scope'
stdin = None

    def execute_java(classpath, main, jvm_options=None, args=None, executor=None,
                     workunit_factory=None, workunit_name=None, workunit_labels=None,
                     cwd=None, workunit_log_config=None, distribution=None,
                     create_synthetic_jar=True, synthetic_jar_dir=None, stdin=None):
      """Executes the java program defined by the classpath and main.
    
      If `workunit_factory` is supplied, does so in the context of a workunit.
    
      :param list classpath: the classpath for the java program
      :param string main: the fully qualified class name of the java program's entry point
      :param list jvm_options: an optional sequence of options for the underlying jvm
      :param list args: an optional sequence of args to pass to the java program
      :param executor: an optional java executor to use to launch the program; defaults to a subprocess
        spawn of the default java distribution
      :param workunit_factory: an optional callable that can produce a workunit context
      :param string workunit_name: an optional name for the work unit; defaults to the main
      :param list workunit_labels: an optional sequence of labels for the work unit
      :param string cwd: optionally set the working directory
      :param WorkUnit.LogConfig workunit_log_config: an optional tuple of options affecting reporting
      :param bool create_synthetic_jar: whether to create a synthentic jar that includes the original
        classpath in its manifest.
      :param string synthetic_jar_dir: an optional directory to store the synthetic jar, if `None`
        a temporary directory will be provided and cleaned up upon process exit.
      :param file stdin: The stdin handle to use: by default None, meaning that stdin will
        not be propagated into the process.
    
      Returns the exit code of the java program.
      Raises `pants.java.Executor.Error` if there was a problem launching java itself.
      """
    
      runner = _get_runner(classpath, main, jvm_options, args, executor, cwd, distribution,
                           create_synthetic_jar, synthetic_jar_dir)
      workunit_name = workunit_name or main
    
      return execute_runner(runner,
                            workunit_factory=workunit_factory,
                            workunit_name=workunit_name,
                            workunit_labels=workunit_labels,
                            workunit_log_config=workunit_log_config,
>                           stdin=stdin)

pants/java/util.py:80: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

runner = <pants.java.nailgun_executor.NailgunExecutor._runner.<locals>.Runner object at 0x7bf4579d97f0>
workunit_factory = <bound method TestContext.new_workunit of <pants_test.base.context_utils.TestContext object at 0x7bf457a26518>>
workunit_name = 'jar-tool'
workunit_labels = ['TOOL', 'NAILGUN', 'TOOL', 'JVM', 'NAILGUN', 'SUPPRESS_LABEL']
workunit_log_config = None, stdin = None

    def execute_runner(runner, workunit_factory=None, workunit_name=None, workunit_labels=None,
                       workunit_log_config=None, stdin=None):
      """Executes the given java runner.
    
      If `workunit_factory` is supplied, does so in the context of a workunit.
    
      :param runner: the java runner to run
      :param workunit_factory: an optional callable that can produce a workunit context
      :param string workunit_name: an optional name for the work unit; defaults to the main
      :param list workunit_labels: an optional sequence of labels for the work unit
      :param WorkUnit.LogConfig workunit_log_config: an optional tuple of task options affecting reporting
      :param file stdin: The stdin handle to use: by default None, meaning that stdin will
        not be propagated into the process.
    
      Returns the exit code of the java runner.
      Raises `pants.java.Executor.Error` if there was a problem launching java itself.
      """
      if not isinstance(runner, Executor.Runner):
        raise ValueError('The runner argument must be a java Executor.Runner instance, '
                         'given {} of type {}'.format(runner, type(runner)))
    
      if workunit_factory is None:
        return runner.run(stdin=stdin)
      else:
        workunit_labels = [
            WorkUnitLabel.TOOL,
            WorkUnitLabel.NAILGUN if isinstance(runner.executor, NailgunExecutor) else WorkUnitLabel.JVM
        ] + (workunit_labels or [])
    
        with workunit_factory(name=workunit_name, labels=workunit_labels,
                              cmd=runner.cmd, log_config=workunit_log_config) as workunit:
          ret = runner.run(stdout=workunit.output('stdout'),
                           stderr=workunit.output('stderr'),
>                          stdin=stdin)

pants/java/util.py:156: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

this = <pants.java.nailgun_executor.NailgunExecutor._runner.<locals>.Runner object at 0x7bf4579d97f0>
stdout = <_pytest.capture.EncodedFile object at 0x7f27af379a20>
stderr = <_pytest.capture.EncodedFile object at 0x7f27af379a20>, stdin = None
cwd = None

    def run(this, stdout=None, stderr=None, stdin=None, cwd=None):
      nailgun = None
      try:
        nailgun = self._get_nailgun_client(jvm_options, classpath, stdout, stderr, stdin)
        logger.debug('Executing via {ng_desc}: {cmd}'.format(ng_desc=nailgun, cmd=this.cmd))
        return nailgun.execute(main, cwd, *args)
      except (NailgunClient.NailgunError, self.InitialNailgunConnectTimedOut) as e:
        self.terminate()
        raise self.Error('Problem launching via {ng_desc} command {main} {args}: {msg}'
                         .format(ng_desc=nailgun or '<no nailgun connection>',
                                 main=main,
                                 args=' '.join(args),
>                                msg=e))
E       pants.java.executor.Executor.Error: Problem launching via NailgunClient(host='127.0.0.1', port=40527, workdir='/tmp/tmpj6awopxx_BUILD_ROOT') command org.pantsbuild.tools.jar.Main -manifest=/tmp/tmpvtfgz4gu/tmp50t1xzsa -files=/tmp/tmpj6awopxx_BUILD_ROOT/.pants.d/pants_test_backend_jvm_tasks_test_jar_task_TestJarBuilderTask/src.java.pants.agents.fake_agent/tmp2x7hyq3o/FakeAgent.class=FakeAgent.class -update=true -compress=true -default_action=SKIP -skip=^META-INF/[^/]+\.SF$,^META-INF/[^/]+\.DSA$,^META-INF/[^/]+\.RSA$,^META-INF/INDEX.LIST$ -policies=^META-INF/services/=CONCAT_TEXT /tmp/tmpudzfjeo9/tmpweynry1d.jar: ('Problem talking to nailgun server (address: 127.0.0.1:40527, remote_pid=<remote PID chunk not yet received!>, remote_pgrp=<remote PGRP chunk not yet received!>): TruncatedHeaderError("Failed to read nailgun chunk header (TruncatedRead(\'Expected 5 bytes before socket shutdown, instead received 0\',)).",)', TruncatedHeaderError("Failed to read nailgun chunk header (TruncatedRead('Expected 5 bytes before socket shutdown, instead received 0',)).",))

pants/java/nailgun_executor.py:155: Error

During handling of the above exception, another exception occurred:

self = <pants_test.backend.jvm.tasks.test_jar_task.JarBuilderTest testMethod=test_agent_manifest>

    def test_agent_manifest(self):
      self.add_to_build_file('src/java/pants/agents', dedent("""
            java_agent(
              name='fake_agent',
              premain='bob',
              agent_class='fred',
              can_redefine=True,
              can_retransform=True,
              can_set_native_method_prefix=True
            )""").strip())
      java_agent = self.target('src/java/pants/agents:fake_agent')
    
      context = self.context(target_roots=[java_agent])
      jar_builder_task = self.prepare_execute(context)
    
      self.add_to_runtime_classpath(context, java_agent, {'FakeAgent.class': '0xCAFEBABE'})
      with self.jarfile() as existing_jarfile:
        with jar_builder_task.open_jar(existing_jarfile) as jar:
          with jar_builder_task.create_jar_builder(jar) as jar_builder:
>           jar_builder.add_target(java_agent)

pants_test/backend/jvm/tasks/test_jar_task.py:268: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
/opt/pyenv/versions/3.6.7/lib/python3.6/contextlib.py:88: in __exit__
    next(self.gen)
pants/backend/jvm/tasks/jar_task.py:360: in open_jar
    if JarTool.global_instance().run(context=self.context, runjava=self.runjava, args=args):
pants/backend/jvm/subsystems/jar_tool.py:33: in run
    WorkUnitLabel.NAILGUN, WorkUnitLabel.SUPPRESS_LABEL])
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

self = <abc.test_TestJarBuilderTask_test_scope object at 0x7bf457bc9550>
classpath = ['/tmp/tmpj6awopxx_BUILD_ROOT/.pants.d/bootstrap_jvm_tools/7c0a62077e8f/ivy/jars/org.pantsbuild/jar-tool/jars/jar-tool...tmpj6awopxx_BUILD_ROOT/.pants.d/bootstrap_jvm_tools/7c0a62077e8f/ivy/jars/org.pantsbuild/args4j/jars/args4j-0.0.9.jar']
main = 'org.pantsbuild.tools.jar.Main', jvm_options = ['-Xmx256m']
args = ['-manifest=/tmp/tmpvtfgz4gu/tmp50t1xzsa', '-files=/tmp/tmpj6awopxx_BUILD_ROOT/.pants.d/pants_test_backend_jvm_tasks_t...t_action=SKIP', '-skip=^META-INF/[^/]+\\.SF$,^META-INF/[^/]+\\.DSA$,^META-INF/[^/]+\\.RSA$,^META-INF/INDEX.LIST$', ...]
workunit_name = 'jar-tool'
workunit_labels = ['TOOL', 'JVM', 'NAILGUN', 'SUPPRESS_LABEL']
workunit_log_config = None, dist = None

    def runjava(self, classpath, main, jvm_options=None, args=None, workunit_name=None,
                workunit_labels=None, workunit_log_config=None, dist=None):
      """Runs the java main using the given classpath and args.
    
        If --execution-strategy=subprocess is specified then the java main is run in a freshly spawned
        subprocess, otherwise a persistent nailgun server dedicated to this Task subclass is used to
        speed up amortized run times.
    
        :API: public
        """
      executor = self.create_java_executor(dist=dist)
    
      # Creating synthetic jar to work around system arg length limit is not necessary
      # when `NailgunExecutor` is used because args are passed through socket, therefore turning off
      # creating synthetic jar if nailgun is used.
      create_synthetic_jar = self.execution_strategy != self.NAILGUN
      try:
        return util.execute_java(classpath=classpath,
                                 main=main,
                                 jvm_options=jvm_options,
                                 args=args,
                                 executor=executor,
                                 workunit_factory=self.context.new_workunit,
                                 workunit_name=workunit_name,
                                 workunit_labels=workunit_labels,
                                 workunit_log_config=workunit_log_config,
                                 create_synthetic_jar=create_synthetic_jar,
                                 synthetic_jar_dir=self._executor_workdir)
      except executor.Error as e:
>       raise TaskError(e)
E       pants.base.exceptions.TaskError: Problem launching via NailgunClient(host='127.0.0.1', port=40527, workdir='/tmp/tmpj6awopxx_BUILD_ROOT') command org.pantsbuild.tools.jar.Main -manifest=/tmp/tmpvtfgz4gu/tmp50t1xzsa -files=/tmp/tmpj6awopxx_BUILD_ROOT/.pants.d/pants_test_backend_jvm_tasks_test_jar_task_TestJarBuilderTask/src.java.pants.agents.fake_agent/tmp2x7hyq3o/FakeAgent.class=FakeAgent.class -update=true -compress=true -default_action=SKIP -skip=^META-INF/[^/]+\.SF$,^META-INF/[^/]+\.DSA$,^META-INF/[^/]+\.RSA$,^META-INF/INDEX.LIST$ -policies=^META-INF/services/=CONCAT_TEXT /tmp/tmpudzfjeo9/tmpweynry1d.jar: ('Problem talking to nailgun server (address: 127.0.0.1:40527, remote_pid=<remote PID chunk not yet received!>, remote_pgrp=<remote PGRP chunk not yet received!>): TruncatedHeaderError("Failed to read nailgun chunk header (TruncatedRead(\'Expected 5 bytes before socket shutdown, instead received 0\',)).",)', TruncatedHeaderError("Failed to read nailgun chunk header (TruncatedRead('Expected 5 bytes before socket shutdown, instead received 0',)).",))

pants/backend/jvm/tasks/nailgun_task.py:123: TaskError
----------------------------- Captured stderr call -----------------------------

Starting workunit cache
.
Starting workunit bootstrap-jar-tool
:: loading settings :: url = jar:file:/home/travis/.ivy2/pants/org.apache.ivy/ivy/jars/ivy-2.4.0.jar!/org/apache/ivy/core/settings/ivysettings.xml
:: resolving dependencies :: internal#42f1fedb02831e5a4cb63034fb17bbb5d4a52248-IvyResolveFingerprintStrategy_76d7a3f2dc89-fetch;working@travis-job-2321f98e-28d0-40b4-844c-429f90eba81f
	confs: [default]
	found org.pantsbuild#jar-tool;0.0.10 in default
	found args4j#args4j;2.32 in default
	found com.google.code.findbugs#jsr305;3.0.0 in default
	found com.google.guava#guava;18.0 in default
	found org.pantsbuild#args4j;0.0.9 in default
:: resolution report :: resolve 298ms :: artifacts dl 17ms
	---------------------------------------------------------------------
	|                  |            modules            ||   artifacts   |
	|       conf       | number| search|dwnlded|evicted|| number|dwnlded|
	---------------------------------------------------------------------
	|      default     |   5   |   0   |   0   |   0   ||   5   |   0   |
	---------------------------------------------------------------------
cachepath output to /tmp/tmpj6awopxx_BUILD_ROOT/.pants.d/bootstrap_jvm_tools/7c0a62077e8f/ivy/42f1fedb02831e5a4cb63034fb17bbb5d4a52248-IvyResolveFingerprintStrategy_76d7a3f2dc89/classpath.raw.tmp.b4296fdbce024a0db1d1fea22815f020

WorkUnit outcome: SUCCESS

Starting workunit cache
.
Starting workunit bootstrap-nailgun-server
:: loading settings :: url = jar:file:/home/travis/.ivy2/pants/org.apache.ivy/ivy/jars/ivy-2.4.0.jar!/org/apache/ivy/core/settings/ivysettings.xml
:: resolving dependencies :: internal#c109bfa74ccd26cbc0a32b4145338ab536124702-IvyResolveFingerprintStrategy_76d7a3f2dc89-fetch;working@travis-job-2321f98e-28d0-40b4-844c-429f90eba81f
	confs: [default]
	found com.martiansoftware#nailgun-server;0.9.1 in default
:: resolution report :: resolve 128ms :: artifacts dl 2ms
	---------------------------------------------------------------------
	|                  |            modules            ||   artifacts   |
	|       conf       | number| search|dwnlded|evicted|| number|dwnlded|
	---------------------------------------------------------------------
	|      default     |   1   |   0   |   0   |   0   ||   1   |   0   |
	---------------------------------------------------------------------
cachepath output to /tmp/tmpj6awopxx_BUILD_ROOT/.pants.d/bootstrap_jvm_tools/7c0a62077e8f/ivy/c109bfa74ccd26cbc0a32b4145338ab536124702-IvyResolveFingerprintStrategy_76d7a3f2dc89/classpath.raw.tmp.4da4c86b42b646b391a526b5268ebd91

WorkUnit outcome: SUCCESS

Starting workunit jar-tool
===================== 2 failed, 11 passed in 50.37 seconds =====================

tests/python/pants_test/backend/jvm/tasks:jar_task stderr:
/home/travis/build/pantsbuild/pants/.pants.d/process-executionx72EnL/pytest-with-requirements.pex/.bootstrap/pex/environment.py:330: PEXWarning: The `pkg_resources` package was loaded from a pex vendored version when declaring namespace packages defined by twitter.common.collections 0.3.11. The twitter.common.collections 0.3.11 distribution should fix its `install_requires` to include `setuptools`
```

--- a/tests/python/pants_test/backend/codegen/wire/java/BUILD
+++ b/tests/python/pants_test/backend/codegen/wire/java/BUILD
@@ -8,15 +8,15 @@ python_tests(
     '3rdparty/python/twitter/commons:twitter.common.collections',
     '3rdparty/python:parameterized',
     'src/python/pants/backend/codegen/wire/java',
-    'src/python/pants/java/jar',
     'src/python/pants/backend/jvm/targets:jvm',
     'src/python/pants/base:exceptions',
     'src/python/pants/base:revision',
     'src/python/pants/base:validation',
     'src/python/pants/build_graph',
+    'src/python/pants/java/jar',
     'src/python/pants/util:contextutil',
+    'tests/python/pants_test/jvm:nailgun_task_test_base',
     'tests/python/pants_test:test_base',
-    'tests/python/pants_test:task_test_base',
   ]
 )
 

--- a/tests/python/pants_test/backend/codegen/wire/java/test_wire_gen.py
+++ b/tests/python/pants_test/backend/codegen/wire/java/test_wire_gen.py
@@ -12,10 +12,10 @@ from pants.backend.codegen.wire.java.wire_gen import WireGen
 from pants.backend.jvm.targets.jar_library import JarLibrary
 from pants.build_graph.register import build_file_aliases as register_core
 from pants.java.jar.jar_dependency import JarDependency
-from pants_test.task_test_base import TaskTestBase
+from pants_test.jvm.nailgun_task_test_base import NailgunTaskTestBase
 
 
-class WireGenTest(TaskTestBase):
+class WireGenTest(NailgunTaskTestBase):
 
   # A bogus target workdir.
   TARGET_WORKDIR = ".pants.d/bogus/workdir"

--- a/tests/python/pants_test/backend/jvm/tasks/BUILD
+++ b/tests/python/pants_test/backend/jvm/tasks/BUILD
@@ -213,14 +213,14 @@ python_tests(
   name = 'detect_duplicates',
   sources = ['test_detect_duplicates.py'],
   dependencies = [
-    'src/python/pants/java/jar',
-    'src/python/pants/backend/jvm/targets:jvm',
+    ':jvm_binary_task_test_base',
     'src/python/pants/backend/jvm/targets:java',
+    'src/python/pants/backend/jvm/targets:jvm',
     'src/python/pants/backend/jvm/tasks:detect_duplicates',
     'src/python/pants/base:exceptions',
+    'src/python/pants/java/jar',
     'src/python/pants/util:contextutil',
     'src/python/pants/util:dirutil',
-    'tests/python/pants_test/jvm:jvm_task_test_base',
   ],
 )
 
@@ -262,21 +262,21 @@ python_tests(
   name = 'ivy_resolve',
   sources = ['test_ivy_resolve.py'],
   dependencies = [
-    '3rdparty/python:future',
     '3rdparty/python/twitter/commons:twitter.common.collections',
-    'src/python/pants/backend/jvm:ivy_utils',
-    'src/python/pants/java/jar',
+    '3rdparty/python:future',
     'src/python/pants/backend/jvm/subsystems:jar_dependency_management',
     'src/python/pants/backend/jvm/targets:java',
     'src/python/pants/backend/jvm/targets:jvm',
     'src/python/pants/backend/jvm/tasks:ivy_resolve',
     'src/python/pants/backend/jvm/tasks:ivy_task_mixin',
+    'src/python/pants/backend/jvm:ivy_utils',
+    'src/python/pants/java/jar',
     'src/python/pants/task',
     'src/python/pants/util:contextutil',
     'tests/python/pants_test/jvm:jvm_tool_task_test_base',
+    'tests/python/pants_test/jvm:nailgun_task_test_base',
     'tests/python/pants_test/subsystem:subsystem_utils',
     'tests/python/pants_test:test_base',
-    'tests/python/pants_test:task_test_base',
   ]
 )
 
@@ -286,21 +286,21 @@ python_tests(
     'test_coursier_resolve.py',
   ],
   dependencies = [
-    '3rdparty/python:future',
     '3rdparty/python/twitter/commons:twitter.common.collections',
+    '3rdparty/python:future',
     '3rdparty/python:mock',
-    'src/python/pants/backend/jvm:ivy_utils',
-    'src/python/pants/java/jar',
     'src/python/pants/backend/jvm/subsystems:jar_dependency_management',
     'src/python/pants/backend/jvm/targets:java',
     'src/python/pants/backend/jvm/targets:jvm',
     'src/python/pants/backend/jvm/tasks:coursier_resolve',
+    'src/python/pants/backend/jvm:ivy_utils',
+    'src/python/pants/java/jar',
     'src/python/pants/task',
     'src/python/pants/util:contextutil',
     'tests/python/pants_test/jvm:jvm_tool_task_test_base',
+    'tests/python/pants_test/jvm:nailgun_task_test_base',
     'tests/python/pants_test/subsystem:subsystem_utils',
     'tests/python/pants_test:test_base',
-    'tests/python/pants_test:task_test_base',
   ]
 )
 
@@ -379,7 +379,7 @@ python_tests(
     'src/python/pants/scm:scm',
     'src/python/pants/util:contextutil',
     'src/python/pants/util:dirutil',
-    'tests/python/pants_test:task_test_base',
+    'tests/python/pants_test/jvm:nailgun_task_test_base',
     'tests/python/pants_test/testutils:mock_logger',
   ],
 )
@@ -481,7 +481,7 @@ python_library(
   dependencies = [
     'src/python/pants/backend/jvm/tasks:classpath_products',
     'src/python/pants/java/jar',
-    'tests/python/pants_test/jvm:jvm_tool_task_test_base',
+    'tests/python/pants_test/jvm:nailgun_task_test_base',
   ]
 )
 

--- a/tests/python/pants_test/backend/jvm/tasks/jvm_binary_task_test_base.py
+++ b/tests/python/pants_test/backend/jvm/tasks/jvm_binary_task_test_base.py
@@ -8,10 +8,10 @@ import os
 
 from pants.backend.jvm.tasks.classpath_products import ClasspathProducts
 from pants.java.jar.jar_dependency_utils import M2Coordinate, ResolvedJar
-from pants_test.jvm.jvm_tool_task_test_base import JvmToolTaskTestBase
+from pants_test.jvm.nailgun_task_test_base import NailgunTaskTestBase
 
 
-class JvmBinaryTaskTestBase(JvmToolTaskTestBase):
+class JvmBinaryTaskTestBase(NailgunTaskTestBase):
   """
   :API: public
   """

--- a/tests/python/pants_test/backend/jvm/tasks/jvm_compile/BUILD
+++ b/tests/python/pants_test/backend/jvm/tasks/jvm_compile/BUILD
@@ -17,9 +17,9 @@ python_tests(
   name = 'jvm_compile',
   sources = ['test_jvm_compile.py'],
   dependencies = [
-    'src/python/pants/backend/jvm/tasks:classpath_products',
     'src/python/pants/backend/jvm/tasks/jvm_compile',
-    'tests/python/pants_test:task_test_base',
+    'src/python/pants/backend/jvm/tasks:classpath_products',
+    'tests/python/pants_test/jvm:nailgun_task_test_base',
   ],
 )
 

--- a/tests/python/pants_test/backend/jvm/tasks/jvm_compile/java/BUILD
+++ b/tests/python/pants_test/backend/jvm/tasks/jvm_compile/java/BUILD
@@ -77,8 +77,8 @@ python_tests(
     'src/python/pants/util:contextutil',
     'src/python/pants/util:dirutil',
     'tests/python/pants_test/java/distribution',
+    'tests/python/pants_test/jvm:nailgun_task_test_base',
     'tests/python/pants_test/subsystem:subsystem_utils',
-    'tests/python/pants_test:task_test_base',
   ],
 )
 

--- a/tests/python/pants_test/backend/jvm/tasks/jvm_compile/java/test_java_compile_settings_partitioning.py
+++ b/tests/python/pants_test/backend/jvm/tasks/jvm_compile/java/test_java_compile_settings_partitioning.py
@@ -19,11 +19,11 @@ from pants.subsystem.subsystem import Subsystem
 from pants.util.memo import memoized_method
 from pants.util.osutil import get_os_name, normalize_os_name
 from pants_test.java.distribution.test_distribution import EXE, distribution
+from pants_test.jvm.nailgun_task_test_base import NailgunTaskTestBase
 from pants_test.subsystem.subsystem_util import init_subsystem
-from pants_test.task_test_base import TaskTestBase
 
 
-class JavaCompileSettingsPartitioningTest(TaskTestBase):
+class JavaCompileSettingsPartitioningTest(NailgunTaskTestBase):
 
   @classmethod
   def task_type(cls):

--- a/tests/python/pants_test/backend/jvm/tasks/jvm_compile/rsc/BUILD
+++ b/tests/python/pants_test/backend/jvm/tasks/jvm_compile/rsc/BUILD
@@ -17,9 +17,9 @@ python_tests(
   name = 'rsc_compile',
   sources = ['test_rsc_compile.py'],
   dependencies = [
-    'src/python/pants/backend/jvm/tasks:classpath_products',
     'src/python/pants/backend/jvm/tasks/jvm_compile',
     'src/python/pants/backend/jvm/tasks/jvm_compile/rsc',
-    'tests/python/pants_test:task_test_base',
+    'src/python/pants/backend/jvm/tasks:classpath_products',
+    'tests/python/pants_test/jvm:nailgun_task_test_base',
   ],
 )

--- a/tests/python/pants_test/backend/jvm/tasks/jvm_compile/rsc/test_rsc_compile.py
+++ b/tests/python/pants_test/backend/jvm/tasks/jvm_compile/rsc/test_rsc_compile.py
@@ -19,8 +19,8 @@ from pants.backend.jvm.tasks.jvm_compile.rsc.rsc_compile import RscCompile, _cre
 from pants.java.jar.jar_dependency import JarDependency
 from pants.option.ranked_value import RankedValue
 from pants.util.contextutil import temporary_dir
+from pants_test.jvm.nailgun_task_test_base import NailgunTaskTestBase
 from pants_test.subsystem.subsystem_util import init_subsystem
-from pants_test.task_test_base import TaskTestBase
 
 
 class LightWeightVTS(object):
@@ -36,7 +36,7 @@ class LightWeightVTS(object):
     pass
 
 
-class RscCompileTest(TaskTestBase):
+class RscCompileTest(NailgunTaskTestBase):
   DEFAULT_CONF = 'default'
 
   @classmethod

--- a/tests/python/pants_test/backend/jvm/tasks/jvm_compile/test_jvm_compile.py
+++ b/tests/python/pants_test/backend/jvm/tasks/jvm_compile/test_jvm_compile.py
@@ -13,14 +13,14 @@ from pants.backend.jvm.tasks.jvm_compile.jvm_compile import JvmCompile
 from pants.backend.jvm.tasks.jvm_compile.zinc.zinc_compile import BaseZincCompile
 from pants.backend.jvm.tasks.nailgun_task import NailgunTaskBase
 from pants.base.build_environment import get_buildroot
-from pants_test.task_test_base import TaskTestBase
+from pants_test.jvm.nailgun_task_test_base import NailgunTaskTestBase
 
 
 class DummyJvmCompile(JvmCompile):
   pass
 
 
-class JvmCompileTest(TaskTestBase):
+class JvmCompileTest(NailgunTaskTestBase):
   DEFAULT_CONF = 'default'
 
   @classmethod
@@ -50,7 +50,7 @@ class JvmCompileTest(TaskTestBase):
       resulting_classpath.get_for_target(target))
 
 
-class BaseZincCompileJDKTest(TaskTestBase):
+class BaseZincCompileJDKTest(NailgunTaskTestBase):
   DEFAULT_CONF = 'default'
   old_cwd = os.getcwd()
 

--- a/tests/python/pants_test/backend/jvm/tasks/test_coursier_resolve.py
+++ b/tests/python/pants_test/backend/jvm/tasks/test_coursier_resolve.py
@@ -22,7 +22,6 @@ from pants.backend.jvm.targets.managed_jar_dependencies import ManagedJarDepende
 from pants.backend.jvm.tasks.classpath_products import ClasspathProducts
 from pants.backend.jvm.tasks.coursier_resolve import (CoursierResolve,
                                                       CoursierResolveFingerprintStrategy)
-from pants.backend.jvm.tasks.nailgun_task import NailgunTask
 from pants.base.exceptions import TaskError
 from pants.java import util
 from pants.java.jar.exclude import Exclude
@@ -30,12 +29,12 @@ from pants.java.jar.jar_dependency import JarDependency
 from pants.task.task import Task
 from pants.util.contextutil import temporary_dir, temporary_file_path
 from pants.util.dirutil import safe_rmtree
-from pants_test.jvm.jvm_tool_task_test_base import JvmToolTaskTestBase
+from pants_test.jvm.nailgun_task_test_base import NailgunTaskTestBase
 from pants_test.subsystem.subsystem_util import init_subsystem
 from pants_test.task_test_base import TaskTestBase
 
 
-class CoursierResolveTest(JvmToolTaskTestBase):
+class CoursierResolveTest(NailgunTaskTestBase):
   """Tests for the class CoursierResolve."""
 
   @classmethod
@@ -44,7 +43,6 @@ class CoursierResolveTest(JvmToolTaskTestBase):
 
   def setUp(self):
     super(CoursierResolveTest, self).setUp()
-    self.set_options(execution_strategy=NailgunTask.ExecutionStrategy.subprocess)
     self.set_options_for_scope('cache.{}'.format(self.options_scope),
                                read_from=None,
                                write_to=None)
@@ -69,22 +67,22 @@ class CoursierResolveTest(JvmToolTaskTestBase):
     compile_classpath = self.resolve([jar_lib, scala_lib])
     self.assertEqual(1, len(compile_classpath.get_for_target(jar_lib)))
     self.assertEqual(0, len(compile_classpath.get_for_target(scala_lib)))
-    
+
   def test_resolve_with_remote_url(self):
     dep_with_url = JarDependency('a', 'b', 'c',
                                  url='http://central.maven.org/maven2/junit/junit/4.12/junit-4.12.jar')
     dep_with_url_lib = self.make_target('//:a', JarLibrary, jars=[dep_with_url])
-    
+
     compile_classpath = self.resolve([dep_with_url_lib])
     # Get paths on compile classpath and assert that it starts with '.../coursier/cache/relative'
     paths = [tup[1] for tup in compile_classpath.get_for_target(dep_with_url_lib)]
     self.assertTrue(any(self._cache_dir_regex('relative').search(path) for path in paths), str(paths))
-  
+
   def test_resolve_with_local_url(self):
     with temporary_file_path(suffix='.jar') as url:
       dep_with_url = JarDependency('commons-lang', 'commons-lang', '2.5', url='file://' + url)
       dep_with_url_lib = self.make_target('//:a', JarLibrary, jars=[dep_with_url])
-      
+
       compile_classpath = self.resolve([dep_with_url_lib])
       # Get paths on compile classpath and assert that it starts with '.../coursier/cache/absolute'
       paths = [tup[1] for tup in compile_classpath.get_for_target(dep_with_url_lib)]

--- a/tests/python/pants_test/backend/jvm/tasks/test_detect_duplicates.py
+++ b/tests/python/pants_test/backend/jvm/tasks/test_detect_duplicates.py
@@ -15,10 +15,10 @@ from pants.java.jar.jar_dependency import JarDependency
 from pants.java.jar.jar_dependency_utils import M2Coordinate, ResolvedJar
 from pants.util.contextutil import open_zip
 from pants.util.dirutil import safe_mkdir, safe_mkdir_for, touch
-from pants_test.jvm.jvm_task_test_base import JvmTaskTestBase
+from pants_test.backend.jvm.tasks.jvm_binary_task_test_base import JvmBinaryTaskTestBase
 
 
-class DuplicateDetectorTest(JvmTaskTestBase):
+class DuplicateDetectorTest(JvmBinaryTaskTestBase):
 
   @classmethod
   def task_type(cls):

--- a/tests/python/pants_test/backend/jvm/tasks/test_ivy_imports.py
+++ b/tests/python/pants_test/backend/jvm/tasks/test_ivy_imports.py
@@ -12,7 +12,6 @@ from pants.backend.jvm.targets.jar_library import JarLibrary
 from pants.backend.jvm.targets.unpacked_jars import UnpackedJars
 from pants.backend.jvm.tasks.ivy_imports import IvyImports
 from pants.backend.jvm.tasks.jar_import_products import JarImportProducts
-from pants.backend.jvm.tasks.nailgun_task import NailgunTask
 from pants.java.jar.jar_dependency import JarDependency
 from pants.java.jar.jar_dependency_utils import M2Coordinate
 from pants.util.contextutil import open_zip, temporary_dir
@@ -53,7 +52,6 @@ class IvyImportsTest(NailgunTaskTestBase):
                                     libraries=[jar_library.address.spec],
                                     include_patterns=['a/b/c/*.proto'])
 
-      self.set_options(execution_strategy=NailgunTask.ExecutionStrategy.subprocess)
       ivy_imports_task = self.execute(self.context(target_roots=[foo_target]))
 
       # Make sure the product is properly populated

--- a/tests/python/pants_test/backend/jvm/tasks/test_ivy_resolve.py
+++ b/tests/python/pants_test/backend/jvm/tasks/test_ivy_resolve.py
@@ -21,14 +21,13 @@ from pants.backend.jvm.targets.jvm_target import JvmTarget
 from pants.backend.jvm.targets.managed_jar_dependencies import ManagedJarDependencies
 from pants.backend.jvm.tasks.ivy_resolve import IvyResolve
 from pants.backend.jvm.tasks.ivy_task_mixin import IvyResolveFingerprintStrategy
-from pants.backend.jvm.tasks.nailgun_task import NailgunTask
 from pants.ivy.bootstrapper import Bootstrapper
 from pants.java.jar.exclude import Exclude
 from pants.java.jar.jar_dependency import JarDependency
 from pants.task.task import Task
 from pants.util.contextutil import temporary_dir, temporary_file_path
 from pants.util.dirutil import safe_delete
-from pants_test.jvm.jvm_tool_task_test_base import JvmToolTaskTestBase
+from pants_test.jvm.nailgun_task_test_base import NailgunTaskTestBase
 from pants_test.subsystem.subsystem_util import init_subsystem
 from pants_test.task_test_base import TaskTestBase, ensure_cached
 
@@ -37,7 +36,7 @@ def strip_workdir(dir, classpath):
   return [(conf, path[len(dir):]) for conf, path in classpath]
 
 
-class IvyResolveTest(JvmToolTaskTestBase):
+class IvyResolveTest(NailgunTaskTestBase):
   """Tests for the class IvyResolve."""
 
   @classmethod
@@ -46,7 +45,6 @@ class IvyResolveTest(JvmToolTaskTestBase):
 
   def setUp(self):
     super(IvyResolveTest, self).setUp()
-    self.set_options(execution_strategy=NailgunTask.ExecutionStrategy.subprocess)
     self.set_options_for_scope('cache.{}'.format(self.options_scope),
                                read_from=None,
                                write_to=None)

--- a/tests/python/pants_test/backend/jvm/tasks/test_ivy_utils.py
+++ b/tests/python/pants_test/backend/jvm/tasks/test_ivy_utils.py
@@ -20,7 +20,6 @@ from pants.backend.jvm.ivy_utils import (FrozenResolution, IvyFetchStep, IvyInfo
 from pants.backend.jvm.register import build_file_aliases as register_jvm
 from pants.backend.jvm.subsystems.jar_dependency_management import JarDependencyManagement
 from pants.backend.jvm.targets.jar_library import JarLibrary
-from pants.backend.jvm.tasks.nailgun_task import NailgunTask
 from pants.base.build_environment import get_buildroot
 from pants.build_graph.register import build_file_aliases as register_core
 from pants.ivy.ivy_subsystem import IvySubsystem
@@ -401,8 +400,7 @@ class IvyUtilsGenerateIvyTest(IvyUtilsTestBase):
 
   def test_missing_ivy_report(self):
     self.set_options_for_scope(IvySubsystem.options_scope,
-                               cache_dir='DOES_NOT_EXIST',
-                               execution_strategy=NailgunTask.ExecutionStrategy.subprocess)
+                               cache_dir='DOES_NOT_EXIST')
 
     with self.assertRaises(IvyUtils.IvyResolveReportError):
       IvyUtils.parse_xml_report('default', IvyUtils.xml_report_path('INVALID_CACHE_DIR',

--- a/tests/python/pants_test/backend/jvm/tasks/test_jar_publish.py
+++ b/tests/python/pants_test/backend/jvm/tasks/test_jar_publish.py
@@ -22,10 +22,10 @@ from pants.build_graph.target import Target
 from pants.scm.scm import Scm
 from pants.util.contextutil import temporary_dir
 from pants.util.dirutil import safe_mkdir, safe_walk
-from pants_test.task_test_base import TaskTestBase
+from pants_test.jvm.nailgun_task_test_base import NailgunTaskTestBase
 
 
-class JarPublishTest(TaskTestBase):
+class JarPublishTest(NailgunTaskTestBase):
 
   @classmethod
   def task_type(cls):
@@ -293,7 +293,7 @@ class FailNTimesTest(unittest.TestCase):
     foo.bar()
 
 
-class JarPublishAuthTest(TaskTestBase):
+class JarPublishAuthTest(NailgunTaskTestBase):
   """Tests for backend jvm JarPublish class"""
 
   def _default_jvm_opts(self):

--- a/tests/python/pants_test/backend/jvm/tasks/test_jar_task.py
+++ b/tests/python/pants_test/backend/jvm/tasks/test_jar_task.py
@@ -59,7 +59,6 @@ class JarTaskTest(BaseJarTaskTest):
   MAX_SUBPROC_ARGS = 50
 
   class TestJarTask(JarTask):
-
     def execute(self):
       pass
 
@@ -234,7 +233,6 @@ class JarTaskTest(BaseJarTaskTest):
 class JarBuilderTest(BaseJarTaskTest):
 
   class TestJarBuilderTask(JarBuilderTask):
-
     def execute(self):
       pass
 

--- a/tests/python/pants_test/java/BUILD
+++ b/tests/python/pants_test/java/BUILD
@@ -81,9 +81,9 @@ python_tests(
   name = 'test_runjava',
   sources = ['test_runjava_synthetic_classpath.py'],
   dependencies = [
+    'src/python/pants/backend/jvm/tasks:nailgun_task',
     'src/python/pants/java:util',
     'src/python/pants/util:dirutil',
-    'tests/python/pants_test/jvm:jvm_tool_task_test_base',
-    'src/python/pants/backend/jvm/tasks:nailgun_task',
+    'tests/python/pants_test/jvm:nailgun_task_test_base',
   ]
 )

--- a/tests/python/pants_test/java/test_runjava_synthetic_classpath.py
+++ b/tests/python/pants_test/java/test_runjava_synthetic_classpath.py
@@ -12,7 +12,7 @@ from pants.java import util
 from pants.net.http.fetcher import Fetcher
 from pants.util.contextutil import temporary_file
 from pants.util.dirutil import safe_concurrent_rename
-from pants_test.jvm.jvm_tool_task_test_base import JvmToolTaskTestBase
+from pants_test.jvm.nailgun_task_test_base import NailgunTaskTestBase
 
 
 class DummyRunJavaTask(NailgunTask):
@@ -20,14 +20,10 @@ class DummyRunJavaTask(NailgunTask):
     pass
 
 
-class SyntheticClasspathTest(JvmToolTaskTestBase):
+class SyntheticClasspathTest(NailgunTaskTestBase):
   @classmethod
   def task_type(cls):
     return DummyRunJavaTask
-
-  def setUp(self):
-    super(SyntheticClasspathTest, self).setUp()
-    self.set_options(execution_strategy=NailgunTask.ExecutionStrategy.subprocess)
 
   def test_execute_java_no_error_weird_path(self):
     """

--- a/tests/python/pants_test/jvm/BUILD
+++ b/tests/python/pants_test/jvm/BUILD
@@ -25,14 +25,15 @@ python_library(
   sources=['jvm_tool_task_test_base.py'],
   dependencies=[
     ':jvm_task_test_base',
-    'src/python/pants/java/jar',
     'src/python/pants/backend/jvm/subsystems:jvm_tool_mixin',
     'src/python/pants/backend/jvm/targets:jvm',
     'src/python/pants/backend/jvm/targets:scala',
     'src/python/pants/backend/jvm/tasks:bootstrap_jvm_tools',
+    'src/python/pants/backend/jvm/tasks:nailgun_task',
     'src/python/pants/base:build_environment',
     'src/python/pants/build_graph',
     'src/python/pants/ivy',
+    'src/python/pants/java/jar',
   ]
 )
 

--- a/tests/python/pants_test/jvm/jvm_tool_task_test_base.py
+++ b/tests/python/pants_test/jvm/jvm_tool_task_test_base.py
@@ -10,6 +10,7 @@ import shutil
 from pants.backend.jvm.register import build_file_aliases
 from pants.backend.jvm.subsystems.jvm_tool_mixin import JvmToolMixin
 from pants.backend.jvm.tasks.bootstrap_jvm_tools import BootstrapJvmTools
+from pants.backend.jvm.tasks.nailgun_task import NailgunTask
 from pants.base.build_environment import get_pants_cachedir
 from pants.build_graph.build_file_aliases import BuildFileAliases
 from pants.build_graph.target import Target
@@ -60,7 +61,9 @@ class JvmToolTaskTestBase(JvmTaskTestBase):
     # TODO: We really need a straightforward way for pants's own tests to get to the enclosing
     # pants instance's options values.
     artifact_caches = [os.path.join(get_pants_cachedir(), 'artifact_cache')]
-    self.set_options_for_scope(bootstrap_scope, jvm_options=['-Xmx128m'])
+    self.set_options_for_scope(bootstrap_scope,
+                               execution_strategy=NailgunTask.ExecutionStrategy.subprocess,
+                               jvm_options=['-Xmx128m'])
     self.set_options_for_scope('cache.{}'.format(bootstrap_scope),
                                read_from=artifact_caches,
                                write_to=artifact_caches)

--- a/tests/python/pants_test/jvm/nailgun_task_test_base.py
+++ b/tests/python/pants_test/jvm/nailgun_task_test_base.py
@@ -4,12 +4,17 @@
 
 from __future__ import absolute_import, division, print_function, unicode_literals
 
-from pants.backend.jvm.tasks.nailgun_task import NailgunProcessGroup, NailgunTask
+from pants.backend.jvm.tasks.nailgun_task import NailgunTask
 from pants_test.jvm.jvm_tool_task_test_base import JvmToolTaskTestBase
 
 
 class NailgunTaskTestBase(JvmToolTaskTestBase):
-  """Prepares an ephemeral test build root that supports nailgun tasks.
+  """Ensures `NailgunTask` tests use subprocess mode to stably test the task under test.
+
+  For subclasses of NailgunTask the nailgun behavior is irrelevant to the code under test and can
+  cause problems in CI environments. As such, disabling nailgunning ensures the test focus is where
+  it needs to be to test the unit.
+
   :API: public
   """
 
@@ -18,12 +23,4 @@ class NailgunTaskTestBase(JvmToolTaskTestBase):
     :API: public
     """
     super(NailgunTaskTestBase, self).setUp()
-    self.set_options(execution_strategy=NailgunTask.ExecutionStrategy.nailgun)
-
-  def tearDown(self):
-    """
-    :API: public
-    """
-    super(NailgunTaskTestBase, self).tearDown()
-    # Kill any nailguns launched in our ephemeral build root.
-    NailgunProcessGroup(metadata_base_dir=self.subprocess_dir).killall()
+    self.set_options(execution_strategy=NailgunTask.ExecutionStrategy.subprocess)

